packages/plugin-core/src/workspace/tutorialInitializer.ts

```diff
@@ -67,9 +68,51 @@ export class TutorialInitializer
     );
   }
 
+  private getAnalyticsPayloadFromDocument(opts: {
+    document: vscode.TextDocument;
+    ws: DWorkspaceV2;
+  }): TutorialNoteViewedPayload {
+    const { document, ws } = opts;
+    const testGroup = AB_TUTORIAL_TEST.getUserGroup(
+      SegmentClient.instance().anonymousId
+    );
+    const fsPath = document.uri.fsPath;
+    const { vaults, wsRoot } = ws;
+    const vault = VaultUtils.getVaultByFilePath({ vaults, wsRoot, fsPath });
+    const note = file2Note(fsPath, vault);
+    const { fname, custom } = note;
+    const { currentStep, totalSteps } = custom;
+    return {
+      testGroup,
+      fname,
+      currentStep,
+      totalSteps,
+    };
+  }
+
   async onWorkspaceOpen(opts: { ws: DWorkspaceV2 }): Promise<void> {
     const ctx = "TutorialInitializer.onWorkspaceOpen";
 
+    // Register a special analytics handler for the tutorial:
+    // This needs to be registered before we open any tutorial note.
+    // Otherwise some events may be lost and not reported properly.
+    const disposable = vscode.window.onDidChangeActiveTextEditor((e) => {
+      const document = e?.document;
+
+      if (document !== undefined) {
+        const payload = this.getAnalyticsPayloadFromDocument({
```

Comment:
I think this throws if the opened document is not in a vault. Should add a `try...catch` block around it.

Reply:
Seems like the extension host is swallowing the error that is thrown when trying to grab the vault of the file, but I wrapped it around a try catch just to make it safe.